File: apps/api/app/routes/inventory.py
# app/routes/inventory.py
from __future__ import annotations

from datetime import datetime, timezone
from decimal import Decimal
from typing import List, Optional
from uuid import UUID, uuid4

from fastapi import APIRouter, Depends, HTTPException, Request
from sqlalchemy import select
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.models.inventory import InventoryItemORM, StockMoveORM
from app.schemas.inventory import (
    InventoryItemCreateIn,
    InventoryItemOut,
    InventoryItemUpdateIn,
    StockMoveCreateIn,
    StockMoveOut,
)

router = APIRouter(tags=["inventory"])


# ============================================================
# utils
# ============================================================

def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


def _get_actor_store_id(request: Request) -> Optional[UUID]:
    user = getattr(request.state, "user", None)
    store_id = getattr(user, "store_id", None)
    if isinstance(store_id, UUID):
        return store_id
    if isinstance(store_id, str):
        try:
            return UUID(store_id)
        except Exception:
            return None
    return None


def _assert_store_scope(store_id: UUID, actor_store_id: Optional[UUID]) -> None:
    if actor_store_id and store_id != actor_store_id:
        raise HTTPException(status_code=404, detail="Not found")


def _require_store_id(request: Request, body_store_id: Optional[UUID]) -> UUID:
    actor_store_id = _get_actor_store_id(request)
    store_id = actor_store_id or body_store_id
    if not store_id:
        raise HTTPException(status_code=400, detail="store_id required")
    return store_id


# ============================================================
# InventoryItem CRUD
# ============================================================

@router.get("/inventory/items", response_model=List[InventoryItemOut])
def list_inventory_items(
    request: Request,
    q: Optional[str] = None,
    limit: int = 200,
    offset: int = 0,
    db: Session = Depends(get_db),
) -> List[InventoryItemOut]:
    actor_store_id = _get_actor_store_id(request)
    if not actor_store_id:
        raise HTTPException(status_code=400, detail="store_id required")

    stmt = select(InventoryItemORM).where(InventoryItemORM.store_id == actor_store_id)

    if q:
        stmt = stmt.where(InventoryItemORM.name.contains(q))

    stmt = stmt.order_by(InventoryItemORM.name.asc()).limit(limit).offset(offset)
    return db.execute(stmt).scalars().all()


@router.post("/inventory/items", response_model=InventoryItemOut)
def create_inventory_item(
    request: Request,
    body: InventoryItemCreateIn,
    db: Session = Depends(get_db),
) -> InventoryItemOut:
    store_id = _require_store_id(request, body.store_id)
    _assert_store_scope(store_id, _get_actor_store_id(request))

    now = _utcnow()

    item = InventoryItemORM(
        id=uuid4(),
        store_id=store_id,
        sku=(body.sku.strip() if body.sku else None),
        name=body.name.strip(),
        unit=(body.unit.strip() if body.unit else None),
        cost_price=Decimal(str(body.cost_price or 0)),
        sale_price=Decimal(str(body.sale_price or 0)),
        qty_on_hand=Decimal(str(body.qty_on_hand or 0)),
        note=body.note,
        created_at=now,
        updated_at=now,
    )
    db.add(item)

    try:
        db.commit()
    except IntegrityError as e:
        db.rollback()
        raise HTTPException(status_code=409, detail=f"IntegrityError: {e}") from e

    db.refresh(item)
    return item


@router.get("/inventory/items/{item_id}", response_model=InventoryItemOut)
def get_inventory_item(
    request: Request,
    item_id: UUID,
    db: Session = Depends(get_db),
) -> InventoryItemOut:
    item = db.get(InventoryItemORM, item_id)
    if not item:
        raise HTTPException(status_code=404, detail="Not found")

    _assert_store_scope(item.store_id, _get_actor_store_id(request))
    return item


@router.put("/inventory/items/{item_id}", response_model=InventoryItemOut)
def update_inventory_item(
    request: Request,
    item_id: UUID,
    body: InventoryItemUpdateIn,
    db: Session = Depends(get_db),
) -> InventoryItemOut:
    item = db.get(InventoryItemORM, item_id)
    if not item:
        raise HTTPException(status_code=404, detail="Not found")

    _assert_store_scope(item.store_id, _get_actor_store_id(request))

    if body.sku is not None:
        item.sku = body.sku.strip() if body.sku else None
    if body.name is not None:
        item.name = body.name.strip()
    if body.unit is not None:
        item.unit = body.unit.strip() if body.unit else None

    if body.cost_price is not None:
        item.cost_price = Decimal(str(body.cost_price))
    if body.sale_price is not None:
        item.sale_price = Decimal(str(body.sale_price))
    if body.qty_on_hand is not None:
        # 手動上書きは事故りやすいが、棚卸UIがない間の暫定として許可
        item.qty_on_hand = Decimal(str(body.qty_on_hand))

    if body.note is not None:
        item.note = body.note

    item.updated_at = _utcnow()

    try:
        db.commit()
    except IntegrityError as e:
        db.rollback()
        raise HTTPException(status_code=409, detail=f"IntegrityError: {e}") from e

    db.refresh(item)
    return item


@router.delete("/inventory/items/{item_id}")
def delete_inventory_item(
    request: Request,
    item_id: UUID,
    db: Session = Depends(get_db),
) -> dict[str, bool]:
    item = db.get(InventoryItemORM, item_id)
    if not item:
        raise HTTPException(status_code=404, detail="Not found")

    _assert_store_scope(item.store_id, _get_actor_store_id(request))

    db.delete(item)
    db.commit()
    return {"deleted": True}


# ============================================================
# Stock moves (manual)
# ============================================================

@router.get("/inventory/moves", response_model=List[StockMoveOut])
def list_stock_moves(
    request: Request,
    item_id: Optional[UUID] = None,
    limit: int = 200,
    offset: int = 0,
    db: Session = Depends(get_db),
) -> List[StockMoveOut]:
    actor_store_id = _get_actor_store_id(request)
    if not actor_store_id:
        raise HTTPException(status_code=400, detail="store_id required")

    stmt = select(StockMoveORM).where(StockMoveORM.store_id == actor_store_id)

    if item_id:
        stmt = stmt.where(StockMoveORM.item_id == item_id)

    stmt = stmt.order_by(StockMoveORM.created_at.desc()).limit(limit).offset(offset)
    return db.execute(stmt).scalars().all()


@router.post("/inventory/moves", response_model=StockMoveOut)
def create_stock_move(
    request: Request,
    body: StockMoveCreateIn,
    db: Session = Depends(get_db),
) -> StockMoveOut:
    store_id = _require_store_id(request, body.store_id)
    _assert_store_scope(store_id, _get_actor_store_id(request))

    item = db.get(InventoryItemORM, body.item_id)
    if not item or item.store_id != store_id:
        raise HTTPException(status_code=400, detail="Invalid item_id")

    now = _utcnow()

    qty = Decimal(str(body.qty))
    unit_cost = Decimal(str(body.unit_cost)) if body.unit_cost is not None else Decimal(str(item.cost_price or 0))

    # qty_on_hand 更新
    if body.move_type == "in":
        item.qty_on_hand = Decimal(str(item.qty_on_hand)) + qty
    elif body.move_type == "out":
        if Decimal(str(item.qty_on_hand)) < qty:
            raise HTTPException(status_code=400, detail="Insufficient stock")
        item.qty_on_hand = Decimal(str(item.qty_on_hand)) - qty
    elif body.move_type == "adjust":
        # adjust は qty を「差分量」として扱う（+/- は許可しないので、運用上は in/out を推奨）
        # ここでは "adjust" の場合は in と同等に加算として扱う（必要なら別設計に）
        item.qty_on_hand = Decimal(str(item.qty_on_hand)) + qty
    else:
        raise HTTPException(status_code=400, detail="Invalid move_type")

    mv = StockMoveORM(
        id=uuid4(),
        store_id=store_id,
        item_id=item.id,
        move_type=body.move_type,
        qty=qty,
        unit_cost=unit_cost,
        ref_type=body.ref_type,
        ref_id=body.ref_id,
        note=body.note,
        created_at=now,
    )
    db.add(mv)

    try:
        db.commit()
    except IntegrityError as e:
        db.rollback()
        raise HTTPException(status_code=409, detail=f"IntegrityError: {e}") from e

    db.refresh(mv)
    return mv

File: apps/web/src/lib/api/inventory.ts
// Inventory domain
import { apiFetch } from "./core";

export type InventoryItem = {
  id: string;
  store_id: string | null;
  sku: string | null;
  name: string | null;
  unit: string | null;
  cost_price: string | null;
  sale_price: string | null;
  qty_on_hand: string | null;
  note: string | null;
  created_at: string | null;
  updated_at: string | null;
};

export type InventoryItemInput = {
  store_id: string;
  sku?: string | null;
  name: string;
  unit?: string | null;
  cost_price?: number | null;
  sale_price?: number | null;
  qty_on_hand?: number | null;
  note?: string | null;
};

export type StockMove = {
  id: string;
  store_id: string | null;
  item_id: string | null;
  qty_delta: string | null;
  reason: string | null;
  ref_type: string | null;
  ref_id: string | null;
  note: string | null;
  created_at: string | null;
  updated_at: string | null;
};

export type CreateStockMoveInput = {
  store_id: string;
  item_id: string;
  qty_delta: number;
  reason?: string | null;
  ref_type?: string | null;
  ref_id?: string | null;
  note?: string | null;
};

type ListWithMeta<T> = {
  items: T[];
  meta?: { limit?: number; offset?: number; total?: number };
};

function toNullableString(v: unknown): string | null {
  if (v === null || v === undefined) return null;
  if (typeof v === "string") return v;
  if (typeof v === "number" || typeof v === "boolean") return String(v);
  return null;
}

function normalizeInventoryItem(raw: unknown): InventoryItem {
  const r = (raw ?? {}) as Record<string, unknown>;
  return {
    id: toNullableString(r.id) ?? "",
    store_id: toNullableString(r.store_id),
    sku: toNullableString(r.sku),
    name: toNullableString(r.name),
    unit: toNullableString(r.unit),
    cost_price: toNullableString(r.cost_price),
    sale_price: toNullableString(r.sale_price),
    qty_on_hand: toNullableString(r.qty_on_hand),
    note: toNullableString(r.note),
    created_at: toNullableString(r.created_at),
    updated_at: toNullableString(r.updated_at),
  };
}

function normalizeStockMove(raw: unknown): StockMove {
  const r = (raw ?? {}) as Record<string, unknown>;
  return {
    id: toNullableString(r.id) ?? "",
    store_id: toNullableString(r.store_id),
    item_id: toNullableString(r.item_id),
    qty_delta: toNullableString(r.qty_delta),
    reason: toNullableString(r.reason),
    ref_type: toNullableString(r.ref_type),
    ref_id: toNullableString(r.ref_id),
    note: toNullableString(r.note),
    created_at: toNullableString(r.created_at),
    updated_at: toNullableString(r.updated_at),
  };
}

function coerceItemsArray<T>(data: unknown, normalize: (x: unknown) => T): T[] {
  if (data && typeof data === "object" && Array.isArray((data as any).items)) {
    return (data as any).items.map(normalize);
  }
  if (Array.isArray(data)) return data.map(normalize);
  return [];
}

export function getCurrentStoreId(): string | null {
  if (typeof window === "undefined") return null;
  try {
    const v = window.localStorage.getItem("store_id");
    return v && v.trim() ? v.trim() : null;
  } catch {
    return null;
  }
}

export async function listInventoryItems(args?: {
  limit?: number;
  offset?: number;
  store_id?: string;
}): Promise<{ items: InventoryItem[]; meta?: ListWithMeta<unknown>["meta"] }> {
  const p = new URLSearchParams();
  if (args?.limit != null) p.set("limit", String(args.limit));
  if (args?.offset != null) p.set("offset", String(args.offset));
  if (args?.store_id) p.set("store_id", args.store_id);
  const qs = p.toString();
  const data = await apiFetch<unknown>(`/api/v1/inventory/items${qs ? `?${qs}` : ""}`, {
    method: "GET",
    auth: true,
  });
  const items = coerceItemsArray(data, normalizeInventoryItem).filter((x) => x.id);
  const meta =
    data && typeof data === "object" && (data as any).meta
      ? ((data as any).meta as any)
      : undefined;
  return { items, meta };
}

export async function getInventoryItem(itemId: string): Promise<InventoryItem> {
  const data = await apiFetch<unknown>(`/api/v1/inventory/items/${encodeURIComponent(itemId)}`, {
    method: "GET",
    auth: true,
  });
  return normalizeInventoryItem(data);
}

export async function createInventoryItem(input: InventoryItemInput): Promise<InventoryItem> {
  const data = await apiFetch<unknown>("/api/v1/inventory/items", {
    method: "POST",
    auth: true,
    body: input,
  });
  return normalizeInventoryItem(data);
}

export async function updateInventoryItem(
  itemId: string,
  input: Partial<InventoryItemInput>
): Promise<InventoryItem> {
  const data = await apiFetch<unknown>(`/api/v1/inventory/items/${encodeURIComponent(itemId)}`, {
    method: "PUT",
    auth: true,
    body: input,
  });
  return normalizeInventoryItem(data);
}

export async function deleteInventoryItem(itemId: string): Promise<void> {
  await apiFetch<void>(`/api/v1/inventory/items/${encodeURIComponent(itemId)}`, {
    method: "DELETE",
    auth: true,
  });
}

export async function listStockMoves(args?: {
  limit?: number;
  offset?: number;
  store_id?: string;
  item_id?: string;
  ref_id?: string;
}): Promise<{ items: StockMove[]; meta?: ListWithMeta<unknown>["meta"] }> {
  const p = new URLSearchParams();
  if (args?.limit != null) p.set("limit", String(args.limit));
  if (args?.offset != null) p.set("offset", String(args.offset));
  if (args?.store_id) p.set("store_id", args.store_id);
  if (args?.item_id) p.set("item_id", args.item_id);
  if (args?.ref_id) p.set("ref_id", args.ref_id);
  const qs = p.toString();
  const data = await apiFetch<unknown>(`/api/v1/inventory/moves${qs ? `?${qs}` : ""}`, {
    method: "GET",
    auth: true,
  });
  const items = coerceItemsArray(data, normalizeStockMove).filter((x) => x.id);
  const meta =
    data && typeof data === "object" && (data as any).meta
      ? ((data as any).meta as any)
      : undefined;
  return { items, meta };
}

export async function createStockMove(input: CreateStockMoveInput): Promise<StockMove> {
  const data = await apiFetch<unknown>("/api/v1/inventory/moves", {
    method: "POST",
    auth: true,
    body: input,
  });
  return normalizeStockMove(data);
}
